radixdlt-engine/src/main/java/com/radixdlt/application/system/scrypt/EpochUpdateConstraintScrypt.java

```diff
@@ -101,66 +92,92 @@
+  public abstract static class ExpectedEpochChecker implements ReducerState {
     private final UpdatingEpoch updatingEpoch;
+    private final long expectedEpoch;
+
+    protected ExpectedEpochChecker(UpdatingEpoch updatingEpoch) {
+      this.updatingEpoch = updatingEpoch;
+      this.expectedEpoch = updatingEpoch.prevEpoch.epoch() + 1;
+    }
+
+    protected byte[] buildExpectedPrefix() {
+      var expectedPrefix = new byte[2 + Long.BYTES];
+      expectedPrefix[0] = 0;
+      expectedPrefix[1] = 1;
+      Longs.copyTo(expectedEpoch, expectedPrefix, 2);
+      return expectedPrefix;
+    }
+
+    protected final void verifyPrefix(IndexedSubstateIterator<?> iterator)
+        throws ProcedureException {
+      iterator.verifyPostTypePrefixEquals(buildExpectedPrefix());
+    }
+
+    public UpdatingEpoch updatingEpoch() {
+      return updatingEpoch;
+    }
+
+    public long expectedEpoch() {
+      return expectedEpoch;
+    }
+  }
+
+  public final class ProcessExittingStake extends ExpectedEpochChecker {
     private final TreeSet<ExitingStake> exitting =
         new TreeSet<>(
             Comparator.comparing(ExitingStake::dataKey, UnsignedBytes.lexicographicalComparator()));
 
     ProcessExittingStake(UpdatingEpoch updatingEpoch) {
-      this.updatingEpoch = updatingEpoch;
+      super(updatingEpoch);
+    }
+
+    protected byte[] buildExpectedPrefix() {
+      var expectedPrefix = new byte[Long.BYTES + 1];
+      expectedPrefix[0] = 0;
+      Longs.copyTo(expectedEpoch(), expectedPrefix, 1);
+      return expectedPrefix;
     }
 
     public ReducerState process(IndexedSubstateIterator<ExitingStake> indexedSubstateIterator)
         throws ProcedureException {
-      var expectedEpoch = updatingEpoch.prevEpoch.epoch() + 1;
-      var expectedPrefix = new byte[Long.BYTES + 1];
-      Longs.copyTo(expectedEpoch, expectedPrefix, 1);
-      indexedSubstateIterator.verifyPostTypePrefixEquals(expectedPrefix);
-      indexedSubstateIterator
-          .iterator()
-          .forEachRemaining(
-              e -> {
-                // Sanity check
-                if (e.epochUnlocked() != expectedEpoch) {
-                  throw new IllegalStateException(
-                      "Invalid shutdown of exitting stake update epoch expected "
-                          + expectedEpoch
-                          + " but was "
-                          + e.epochUnlocked());
-                }
-                exitting.add(e);
-              });
+      verifyPrefix(indexedSubstateIterator);
+
+      indexedSubstateIterator.forEachRemaining(
+          exitingStake -> exitting.add(validateExittingStake(exitingStake)));
       return next();
     }
 
+    // Sanity check
+    private ExitingStake validateExittingStake(ExitingStake exitingStake) {
+      if (exitingStake.epochUnlocked() != expectedEpoch()) {
+        throw new IllegalStateException(
+            "Invalid shutdown of exitting stake update epoch expected "
```

Comment:
I would use `String.format` here.

Reply:
This is original code which I see no point to touch given how rarely (if ever) it is executed.